src/univariate/discrete/categorical.jl

```diff
@@ -52,6 +52,15 @@ ncategories(d::Categorical) = support(d).stop
 params(d::Categorical{P,Ps}) where {P<:Real, Ps<:AbstractVector{P}} = (probs(d),)
 partype(::Categorical{T}) where {T<:Real} = T
 
+function Base.isapprox(c1::C, c2::C; kwargs...) where {C<:Categorical}
```

Comment:
Wouldn't it better to define this method as
```suggestion
function Base.isapprox(c1::Categorical, c2::Categorical; kwargs...) 
```
There is a big difference in two definitions, as the latter dispatches on
```julia
Categorical([ 1.0, 0.0 ]) ≈ Categorical([ 1f0, 0f0 ]) # Note `Float64` and `Float32`
```

I actually checked this on master and the behaviour is interesting. It means that `Float64 ≈ Float32` dispatches on something else.
```
Categorical([ 1.0, 0.0 ]) ≈ Categorical([ 1.0, 0.0 ])
```

```julia
julia> Categorical([ 1.0, 0.0 ]) ≈ Categorical([ 1f0, 0f0 ])
true

julia> Categorical([ 1.0, 0.0 ]) ≈ Categorical([ 1.0, 0.0 ])
ERROR: ArgumentError: step cannot be zero
Stacktrace:
  [1] steprange_last(start::Int64, step::Int64, stop::Int64)
    @ Base ./range.jl:221
  [2] StepRange
    @ ./range.jl:208 [inlined]
  [3] _rangestyle
    @ ./range.jl:118 [inlined]
  [4] _range
    @ ./range.jl:116 [inlined]
  [5] #range#57
    @ ./range.jl:91 [inlined]
  [6] -
    @ ./range.jl:1107 [inlined]
  [7] isapprox(x::Base.OneTo{Int64}, y::Base.OneTo{Int64}; atol::Int64, rtol::Int64, nans::Bool, norm::typeof(LinearAlgebra.norm))
    @ LinearAlgebra /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/generic.jl:1657
  [8] isapprox
    @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/generic.jl:1657 [inlined]
  [9] isapprox(c1::Categorical{Float64, Vector{Float64}}, c2::Categorical{Float64, Vector{Float64}})
    @ Distributions /private/tmp/Distributions.jl/src/univariate/discrete/discretenonparametric.jl:70
 [10] top-level scope
    @ REPL[4]:1
```

Reply:
Good catch!

The reason for this different behavior is that the fallback in src/common.jl supports arguments of different types (but if the `nameof` the types - `DiscreteNonParametric` here - or the field names are not identical, `false` is returned). And the default definition there compares fields with `===` first before evaluating `==` or `isapprox`, and hence in this example for different the check of `Base.OneTo` supports with `isapprox` is never performed.

Since the fallback definition supports arguments of different types, I guess it would be better to support arguments of non-identical types here as well. I changed it.